### ISSUE TITLE
Using function like goto_map() in sample project causes MetSys to change maps twice

### ISSUE THREAD

**Issue by dehboxturtle** [KoBeWi/Metroidvania-System#25]
I used your game script in the sample project as a base for when I started my game and I noticed some interesting behavior if you use `goto_map() `manually (such as in the portal block). Calling it will correctly load the new map, however since the user is responsible for setting the players position in the `_process` function, once the map loads it will trigger the metsys signal `room_changed` and end up calling `goto_map()` again.

I believe you can reproduce this in the sample project as well by putting a break point in the `goto_map()` method and entering the portal. I believe this is undesirable behavior because it isn't obvious it will happen and it results in multiple `RoomInstance` nodes ending up in the scene tree. In my project I reproduced this by calling `goto_map()` from a portal like block that teleports you to a distance room in the same z layer.

I was able to workaround this in my project by simply adding a `teleport_to_map()` function that is the same as `goto_map()` but just sets a variable so that it knows to do nothing on the next `room_changed` signal. It would be nice if MetSys provided a function for teleportation to avoid this situation though since I believe it is not obvious that this happens and could confuse users, particularly if they change state when they load their scenes as the logic will get triggered twice.

Happy to help contribute a change to metsys or the sample project if you agree this should be changed and have a desired approach in mind.

**Comment by KoBeWi** [comment]
The sample project circumvents this by using player's `event` variable. I checked and the portal works correctly, however the game hangs when I put a breakpoint 🙃
The problem was caused by this line in Portal.gd
```GDScript
get_tree().create_timer(0.05).timeout.connect(body.set.bind(&"event", false))
```
It's fixed by increasing the timer to 0.2.

I'm currently reworking some functions and they will be provided in a reusable way rather than forcing you to extract them from sample project, maybe I will come up with something more reliable.

**Comment by dehboxturtle** [comment]
I don't actually see any code in the sample project where the game object checks if the player is in the event state, that seems to just be tied to the players movement?

I took a look at the sample project in the repo and it is different from when I originally checked it out though. I think a commit or something might have gotten dropped somewhere as well because there are references in scripts to functions that no longer exist (such as portal.gd referring to the `goto_map()` function). So we're probably just looking at different things.

Here is a link to a video showing the problem I was encountering. The portal itself is working fine, but you'll see that as I progress through the code I end up in the `_ready()` function of my level twice because of the double loading thing I mentioned. 
https://www.dropbox.com/scl/fi/j5txwr808cnl9nfajrkno/demo_of_double_load.mp4?rlkey=dxgtfan2blxsyc5b2pps6eqv0&dl=0

The issue in my project before I fixed it is that when someone loads in a new map and adds it as a child to the game, that doesn't result in the player's position in room coordinates being updated within metsys. So the whole thing will load in, then on the first frame when you call `Metsys.setPlayerPosition()` it will see that the player is in a different room from the previous frame and emit the `room_changed` signal, causing the level to be loaded in again.

I think as long as after the updates you are making Metsys provides a way for the player to change levels manually that also updates the internal state of metsys so that the players room coordinates get changed then it should be much harder for people to run into a situation like this.

**Comment by KoBeWi** [comment]
Hmm I can't reproduce that problem 🤔
The `event` is used only in portal script. The code on the repo has changed significantly today, because I refactored some parts.

Still, I downloaded 1.1 build and when entering the portal, `on_room_changed()` is not called at all. Though the portals in sample project only change the layer. I guess it's different in your case?

**Comment by dehboxturtle** [comment]
Ya in my case I'm moving the player to a distant room in the same layer. I guess it must just be something with how I've modified the game singleton for my game maybe that was causing it then? 

Just to be sure, once you've finished pushing all your changes to things I'll try to update metsys in my game and also do a clean pull of the sample project to see if I can repro it in just the sample project again. Thanks for taking a look!

**Comment by KoBeWi** [comment]
Fixed in d6cf9d04a06a82c545f93fbd251ee42de58feb56
In the old code you can add a similar condition in `on_room_changed()` in Game.gd. I'd avoid upgrading the addon until I officially release the new version.
